File: AgentUp.Desktop/Features/FakeServer/Providers/FakeGitProvider.cs
using System.Text.Json.Nodes;
using AgentUp.Desktop.Features.FakeServer.Models;

namespace AgentUp.Desktop.Features.FakeServer.Providers;

public static class FakeGitProvider
{
    public static JsonObject Head(JsonObject git)
    {
        var changes = ChangesObject(git);
        return new JsonObject
        {
            ["branch"] = changes["branch"]?.DeepClone() ?? "main",
            ["localBranches"] = changes["localBranches"]?.DeepClone() ?? new JsonArray("main"),
            ["remoteBranches"] = changes["remoteBranches"]?.DeepClone() ?? new JsonArray(new JsonObject { ["remote"] = "origin", ["name"] = "main" }),
            ["upstream"] = changes["upstream"]?.DeepClone() ?? "origin/main",
            ["ahead"] = changes["ahead"]?.DeepClone() ?? 0,
            ["behind"] = changes["behind"]?.DeepClone() ?? 0,
            ["commit"] = changes["commit"]?.DeepClone()
        };
    }

    public static JsonObject Commit(JsonObject git, IReadOnlyList<string> files, string message)
    {
        var selected = files.Where(path => FindFile(git, path) is not null).ToArray();
        if (selected.Length == 0)
            return Result(succeeded: false, "Select at least one file to commit.");

        var commit = NextCommit(git);
        RemoveFiles(git, selected);
        SetInt(ChangesObject(git), "ahead", IntValue(ChangesObject(git), "ahead") + 1);
        ChangesObject(git)["commit"] = commit[..Math.Min(7, commit.Length)];
        PrependLog(git, commit, message.Trim().Length == 0 ? "chore: update harbor shop" : message.Trim(), "Demo", ["HEAD", Branch(git)]);
        return Result(succeeded: true, commit: commit);
    }

    public static JsonObject Discard(JsonObject git, IReadOnlyList<string> files)
    {
        RemoveFiles(git, files);
        return Result(succeeded: true, head: Head(git));
    }

    public static JsonObject Fetch(JsonObject git)
    {
        if (IntValue(ChangesObject(git), "behind") == 0)
            SetInt(ChangesObject(git), "behind", 1);
        return Result(succeeded: true, head: Head(git));
    }

    public static JsonObject Pull(JsonObject git)
    {
        if (IntValue(ChangesObject(git), "behind") > 0)
        {
            var commit = NextCommit(git);
            SetInt(ChangesObject(git), "behind", 0);
            ChangesObject(git)["commit"] = commit[..Math.Min(7, commit.Length)];
            PrependLog(git, commit, "chore(storefront): restock the harbor mug", "origin", ["origin/main"]);
        }

        return Result(succeeded: true, head: Head(git));
    }

    public static JsonObject Push(JsonObject git)
    {
        SetInt(ChangesObject(git), "ahead", 0);
        return Result(succeeded: true, head: Head(git));
    }

    public static JsonObject SwitchBranch(JsonObject git, string name, bool create)
    {
        var branch = name.Trim();
        if (branch.Length == 0)
            return Result(succeeded: false, "A branch name is required.");
        var locals = LocalBranches(git);
        if (create && !locals.Contains(branch, StringComparer.Ordinal))
            locals.Add(branch);
        if (!locals.Contains(branch, StringComparer.Ordinal))
            return Result(succeeded: false, $"Branch {branch} does not exist.");
        WriteLocalBranches(git, locals);
        ChangesObject(git)["branch"] = branch;
        return Result(succeeded: true, head: Head(git));
    }

    public static JsonObject CheckoutRemote(JsonObject git, string name)
    {
        var raw = name.Trim();
        if (raw.Length == 0)
            return Result(succeeded: false, "A branch name is required.");
        var local = raw.Contains('/', StringComparison.Ordinal) ? raw[(raw.LastIndexOf('/') + 1)..] : raw;
        var locals = LocalBranches(git);
        if (!locals.Contains(local, StringComparer.Ordinal))
            locals.Add(local);
        WriteLocalBranches(git, locals);
        ChangesObject(git)["branch"] = local;
        return Result(succeeded: true, head: Head(git));
    }

    public static string AddAgentFile(JsonObject git)
    {
        var sequence = IntValue(git, "nextAgentFile");
        if (sequence == 0)
            sequence = 1;
        git["nextAgentFile"] = sequence + 1;
        var suffix = sequence == 1 ? "" : sequence.ToString();
        var path = $"apps/storefront/PromoBanner{suffix}.tsx";
        var diff = $"--- /dev/null\n+++ b/{path}\n@@ -0,0 +1,5 @@\n+export function PromoBanner{suffix}() {{\n+  return <aside>Weekly harbor special</aside>;\n+}}\n";
        UpsertFile(git, path, "Added", diff);
        return path;
    }

    public static JsonNode? Diff(JsonObject git, string path)
        => git["diffs"]?[path];

    private static void UpsertFile(JsonObject git, string path, string status, string diff)
    {
        var files = Flatten(git).Where(file => file.Path != path).ToList();
        files.Add(new FakeGitFile(path, Path.GetFileName(path), status, diff));
        WriteFiles(git, files);
        if (git["diffs"] is not JsonObject diffs)
        {
            diffs = [];
            git["diffs"] = diffs;
        }

        diffs[path] = new JsonObject
        {
            ["path"] = path,
            ["status"] = status,
            ["isBinary"] = false,
            ["diff"] = diff
        };
        var unassigned = Unassigned(git);
        if (!unassigned.Contains(path, StringComparer.Ordinal))
            unassigned.Add(path);
        WriteUnassigned(git, unassigned);
        ChangesObject(git)["fileCount"] = files.Count;
    }

    private static void RemoveFiles(JsonObject git, IEnumerable<string> paths)
    {
        var remove = paths.ToHashSet(StringComparer.Ordinal);
        var files = Flatten(git).Where(file => !remove.Contains(file.Path)).ToList();
        WriteFiles(git, files);
        if (git["diffs"] is JsonObject diffs)
        {
            foreach (var path in remove)
                diffs.Remove(path);
        }

        WriteUnassigned(git, Unassigned(git).Where(path => !remove.Contains(path)).ToList());
        ChangesObject(git)["fileCount"] = files.Count;
    }

    private static List<FakeGitFile> Flatten(JsonObject git)
    {
        var files = new List<FakeGitFile>();
        Collect(ChangesObject(git)["root"] as JsonObject, git["diffs"] as JsonObject, files);
        return files;
    }

    private static void Collect(JsonObject? directory, JsonObject? diffs, List<FakeGitFile> files)
    {
        if (directory is null)
            return;
        foreach (var child in (directory["directories"] as JsonArray ?? []).OfType<JsonObject>())
            Collect(child, diffs, files);
        files.AddRange((directory["files"] as JsonArray ?? [])
            .OfType<JsonObject>()
            .Select(file => (
                Path: file["path"]?.GetValue<string>(),
                Name: file["name"]?.GetValue<string>(),
                Status: file["status"]?.GetValue<string>() ?? "Modified"))
            .Where(file => !string.IsNullOrWhiteSpace(file.Path))
            .Select(file => new FakeGitFile(
                file.Path!,
                file.Name ?? Path.GetFileName(file.Path!),
                string.Equals(file.Status, "modified", StringComparison.OrdinalIgnoreCase) ? "Modified" : file.Status,
                diffs?[file.Path!]?["diff"]?.GetValue<string>() ?? $"--- a/{file.Path}\n+++ b/{file.Path}\n")));
    }

    private static void WriteFiles(JsonObject git, List<FakeGitFile> files)
    {
        var root = new JsonObject
        {
            ["name"] = "",
            ["path"] = "",
            ["directories"] = new JsonArray(),
            ["files"] = new JsonArray()
        };
        foreach (var file in files)
        {
            var parts = file.Path.Split('/', StringSplitOptions.RemoveEmptyEntries);
            var directory = root;
            for (var index = 0; index < parts.Length - 1; index++)
            {
                var path = string.Join('/', parts.Take(index + 1));
                var directories = (JsonArray)directory["directories"]!;
                var child = directories.OfType<JsonObject>()
                    .FirstOrDefault(item => item["path"]?.GetValue<string>() == path);
                if (child is null)
                {
                    child = new JsonObject
                    {
                        ["name"] = parts[index],
                        ["path"] = path,
                        ["directories"] = new JsonArray(),
                        ["files"] = new JsonArray()
                    };
                    directories.Add(child);
                }

                directory = child;
            }

            ((JsonArray)directory["files"]!).Add(new JsonObject
            {
                ["name"] = file.Name,
                ["path"] = file.Path,
                ["status"] = file.Status
            });
        }

        ChangesObject(git)["root"] = root;
    }

    private static FakeGitFile? FindFile(JsonObject git, string path)
        => Flatten(git).FirstOrDefault(file => file.Path == path);

    private static JsonObject ChangesObject(JsonObject git)
        => git["changes"] as JsonObject ?? throw new InvalidOperationException("The fake git node is missing changes.");

    private static string Branch(JsonObject git)
        => ChangesObject(git)["branch"]?.GetValue<string>() ?? "main";

    private static List<string> LocalBranches(JsonObject git)
        => (ChangesObject(git)["localBranches"] as JsonArray ?? [])
            .Select(item => item?.GetValue<string>())
            .Where(item => !string.IsNullOrWhiteSpace(item))
            .Cast<string>()
            .ToList();

    private static void WriteLocalBranches(JsonObject git, IEnumerable<string> branches)
        => ChangesObject(git)["localBranches"] = new JsonArray(branches.Select(branch => JsonValue.Create(branch)).ToArray());

    private static List<string> Unassigned(JsonObject git)
        => (git["queue"]?["unassignedFiles"] as JsonArray ?? [])
            .Select(item => item?.GetValue<string>())
            .Where(item => !string.IsNullOrWhiteSpace(item))
            .Cast<string>()
            .ToList();

    private static void WriteUnassigned(JsonObject git, IEnumerable<string> files)
    {
        if (git["queue"] is not JsonObject queue)
        {
            queue = [];
            git["queue"] = queue;
        }

        queue["unassignedFiles"] = new JsonArray(files.Select(file => JsonValue.Create(file)).ToArray());
    }

    private static void PrependLog(JsonObject git, string commit, string subject, string author, string[] refs)
    {
        if (git["log"] is not JsonObject log)
        {
            log = new JsonObject { ["commits"] = new JsonArray() };
            git["log"] = log;
        }

        var commits = log["commits"] as JsonArray ?? [];
        var parents = new JsonArray();
        if (commits.Count > 0 && commits[0]?["id"]?.GetValue<string>() is { } parent)
            parents.Add(parent);
        foreach (var existing in commits.OfType<JsonObject>().Where(item => item["refs"] is JsonArray))
        {
            var current = (JsonArray)existing["refs"]!;
            existing["refs"] = new JsonArray(current.Where(item => item?.GetValue<string>() != "HEAD").Select(item => item!.DeepClone()).ToArray());
        }

        commits.Insert(0, new JsonObject
        {
            ["id"] = commit,
            ["shortId"] = commit[..Math.Min(7, commit.Length)],
            ["parents"] = parents,
            ["subject"] = subject,
            ["author"] = author,
            ["timestamp"] = DateTimeOffset.UtcNow.ToString("O"),
            ["refs"] = new JsonArray(refs.Select(item => JsonValue.Create(item)).ToArray())
        });
        log["commits"] = commits;
    }

    private static string NextCommit(JsonObject git)
    {
        var next = IntValue(git, "nextCommit");
        if (next == 0)
            next = 1;
        git["nextCommit"] = next + 1;
        return $"f4ke{next:D8}";
    }

    private static int IntValue(JsonObject node, string name)
    {
        if (node[name] is not JsonValue json)
            return 0;
        if (json.TryGetValue<int>(out var number))
            return number;
        return json.TryGetValue<long>(out var longer) ? (int)longer : 0;
    }

    private static void SetInt(JsonObject node, string name, int value)
        => node[name] = value;

    private static JsonObject Result(bool succeeded, string? error = null, string? commit = null, JsonObject? head = null)
    {
        var payload = new JsonObject
        {
            ["found"] = true,
            ["succeeded"] = succeeded,
            ["error"] = error,
            ["commit"] = commit
        };
        if (head is not null)
            payload["head"] = head;
        return payload;
    }

}

File: AgentUp.Mobile/src/features/fake-server/providers/FakeGitProvider.ts
import type { GitChangeDirectory, GitChangeStatus, GitHeadState, GitLogCommit } from '@/features/git/models/GitChanges';

export type FakeGitFile = {
  path: string;
  status: GitChangeStatus;
  diff: string;
};

export type FakeGitState = {
  workspaceId: string;
  branch: string;
  commit: string;
  ahead: number;
  behind: number;
  localBranches: string[];
  remoteBranches: Array<{ remote: string; name: string }>;
  upstream: string | null;
  files: FakeGitFile[];
  log: GitLogCommit[];
  unassignedFiles: string[];
  queueWorktreePath: string | null;
  baseCommit: string;
  tipCommit: string;
  generation: number;
  entries: unknown[];
  nextCommit: number;
  nextAgentFile: number;
  incomingCommit: string | null;
  incomingPulled: boolean;
};

export type FakeGitMutation = {
  found: boolean;
  succeeded: boolean;
  error: string | null;
  commit?: string | null;
  head?: GitHeadState | null;
};

const statuses: GitChangeStatus[] = ['Modified', 'Added', 'Deleted', 'Renamed', 'Untracked', 'Conflicted'];

export function normalizeGitStatus(value: string | null | undefined): GitChangeStatus {
  if (!value) return 'Modified';
  const match = statuses.find(status => status.toLowerCase() === value.toLowerCase());
  return match ?? 'Modified';
}

export function loadFakeGitState(workspaceId: string, node: Record<string, unknown> | undefined): FakeGitState | null {
  if (!node) return null;
  const changes = asRecord(node.changes);
  const diffs = asRecord(node.diffs) ?? {};
  const log = asRecord(node.log);
  const queue = asRecord(node.queue);
  const files = collectFiles(asRecord(changes?.root), diffs);
  const commits = Array.isArray(log?.commits) ? log.commits.map(asLogCommit).filter((item): item is GitLogCommit => !!item) : [];
  const unassigned = Array.isArray(queue?.unassignedFiles)
    ? queue.unassignedFiles.filter((item): item is string => typeof item === 'string')
    : files.map(file => file.path);
  return {
    workspaceId,
    branch: stringValue(changes?.branch) ?? 'main',
    commit: stringValue(changes?.commit) ?? 'a1b2c3d',
    ahead: numberValue(changes?.ahead) ?? 0,
    behind: numberValue(changes?.behind) ?? 0,
    localBranches: stringArray(changes?.localBranches) ?? ['main'],
    remoteBranches: remoteBranches(changes?.remoteBranches),
    upstream: stringValue(changes?.upstream) ?? 'origin/main',
    files,
    log: commits,
    unassignedFiles: unassigned,
    queueWorktreePath: stringValue(queue?.queueWorktreePath),
    baseCommit: stringValue(queue?.baseCommit) ?? commits[0]?.id ?? 'a1b2c3d4e5f6',
    tipCommit: stringValue(queue?.tipCommit) ?? commits[0]?.id ?? 'a1b2c3d4e5f6',
    generation: numberValue(queue?.generation) ?? 0,
    entries: Array.isArray(queue?.entries) ? queue.entries : [],
    nextCommit: 1,
    nextAgentFile: 1,
    incomingCommit: stringValue(node.incomingCommit),
    incomingPulled: node.incomingPulled === true,
  };
}

export function gitChangesPayload(git: FakeGitState) {
  return {
    workspaceId: git.workspaceId,
    branch: git.branch,
    fileCount: git.files.length,
    commit: git.commit,
    ahead: git.ahead,
    behind: git.behind,
    localBranches: git.localBranches,
    remoteBranches: git.remoteBranches,
    upstream: git.upstream,
    root: buildChangeTree(git.files),
  };
}

export function gitHeadPayload(git: FakeGitState): GitHeadState {
  return {
    branch: git.branch,
    localBranches: [...git.localBranches],
    remoteBranches: [...git.remoteBranches],
    upstream: git.upstream,
    ahead: git.ahead,
    behind: git.behind,
    commit: git.commit,
  };
}

export function gitLogPayload(git: FakeGitState) {
  return { commits: git.log, hasMore: false };
}

export function gitQueuePayload(git: FakeGitState) {
  return {
    entries: git.entries,
    unassignedFiles: git.unassignedFiles,
    queueWorktreePath: git.queueWorktreePath,
    baseCommit: git.baseCommit,
    tipCommit: git.tipCommit,
    generation: git.generation,
  };
}

export function gitDiffPayload(git: FakeGitState, path: string): FakeGitFile | null {
  return git.files.find(file => file.path === path) ?? null;
}

export function commitGitFiles(git: FakeGitState, files: string[], message: string): FakeGitMutation {
  const selected = files.filter(path => git.files.some(file => file.path === path));
  if (selected.length === 0) return { found: true, succeeded: false, error: 'Select at least one file to commit.', commit: null };
  const commit = nextCommitId(git);
  git.files = git.files.filter(file => !selected.includes(file.path));
  git.unassignedFiles = git.unassignedFiles.filter(path => !selected.includes(path));
  git.ahead += 1;
  git.commit = commit.slice(0, 7);
  git.tipCommit = commit;
  git.log = [
    {
      id: commit,
      shortId: commit.slice(0, 7),
      parents: git.log[0] ? [git.log[0].id] : [],
      subject: message.trim() || 'chore: update harbor shop',
      author: 'Demo',
      timestamp: new Date().toISOString(),
      refs: ['HEAD', git.branch],
    },
    ...git.log.map(entry => ({ ...entry, refs: entry.refs.filter(ref => ref !== 'HEAD') })),
  ];
  fetchGitRemote(git);
  return { found: true, succeeded: true, error: null, commit, head: gitHeadPayload(git) };
}

export function discardGitFiles(git: FakeGitState, files: string[]): FakeGitMutation {
  git.files = git.files.filter(file => !files.includes(file.path));
  git.unassignedFiles = git.unassignedFiles.filter(path => !files.includes(path));
  return { found: true, succeeded: true, error: null, head: gitHeadPayload(git) };
}

export function fetchGitRemote(git: FakeGitState): FakeGitMutation {
  if (git.incomingPulled) return syncResult(git);
  if (git.behind === 0) git.behind = 1;
  ensureIncomingCommit(git);
  return syncResult(git);
}

export function pullGitRemote(git: FakeGitState): FakeGitMutation {
  fetchGitRemote(git);
  if (git.behind === 0 || git.incomingPulled) return syncResult(git);
  const commit = ensureIncomingCommit(git);
  git.behind = 0;
  git.incomingPulled = true;
  git.commit = commit.slice(0, 7);
  git.tipCommit = commit;
  git.log = [
    {
      id: commit,
      shortId: commit.slice(0, 7),
      parents: git.log[0] ? [git.log[0].id] : [],
      subject: 'chore(storefront): restock the harbor mug',
      author: 'origin',
      timestamp: new Date().toISOString(),
      refs: ['HEAD', git.branch, git.upstream ?? 'origin/main'],
    },
    ...git.log.map(entry => ({
      ...entry,
      refs: entry.refs.filter(ref => ref !== 'HEAD' && ref !== git.branch),
    })),
  ];
  return syncResult(git);
}

export function pushGitRemote(git: FakeGitState): FakeGitMutation {
  git.ahead = 0;
  return syncResult(git);
}

export function switchGitBranch(git: FakeGitState, name: string, create: boolean): FakeGitMutation {
  const branch = name.trim();
  if (!branch) return { found: true, succeeded: false, error: 'A branch name is required.' };
  if (create && !git.localBranches.includes(branch)) git.localBranches = [...git.localBranches, branch];
  if (!git.localBranches.includes(branch)) return { found: true, succeeded: false, error: `Branch ${branch} does not exist.` };
  git.branch = branch;
  return { found: true, succeeded: true, error: null, head: gitHeadPayload(git) };
}

export function checkoutGitRemote(git: FakeGitState, name: string): FakeGitMutation {
  const raw = name.trim();
  if (!raw) return { found: true, succeeded: false, error: 'A branch name is required.' };
  const local = raw.includes('/') ? raw.slice(raw.lastIndexOf('/') + 1) : raw;
  if (!git.localBranches.includes(local)) git.localBranches = [...git.localBranches, local];
  git.branch = local;
  return { found: true, succeeded: true, error: null, head: gitHeadPayload(git) };
}

export function addAgentWorkingTreeFile(git: FakeGitState): FakeGitFile {
  const suffix = git.nextAgentFile === 1 ? '' : String(git.nextAgentFile);
  git.nextAgentFile += 1;
  const path = `apps/storefront/PromoBanner${suffix}.tsx`;
  const file: FakeGitFile = {
    path,
    status: 'Added',
    diff: `--- /dev/null\n+++ b/${path}\n@@ -0,0 +1,5 @@\n+export function PromoBanner${suffix}() {\n+  return <aside>Weekly harbor special</aside>;\n+}\n`,
  };
  git.files = [...git.files.filter(item => item.path !== path), file];
  if (!git.unassignedFiles.includes(path)) git.unassignedFiles = [...git.unassignedFiles, path];
  return file;
}

export function buildChangeTree(files: FakeGitFile[]): GitChangeDirectory {
  const root: GitChangeDirectory = { name: '', path: '', directories: [], files: [] };
  for (const file of files) {
    const parts = file.path.split('/').filter(Boolean);
    let directory = root;
    for (let index = 0; index < parts.length - 1; index += 1) {
      const path = parts.slice(0, index + 1).join('/');
      let child = directory.directories.find(item => item.path === path);
      if (!child) {
        child = { name: parts[index], path, directories: [], files: [] };
        directory.directories.push(child);
      }
      directory = child;
    }
    const name = parts.at(-1) ?? file.path;
    directory.files.push({ name, path: file.path, status: file.status });
  }
  return root;
}

function collectFiles(root: Record<string, unknown> | undefined, diffs: Record<string, unknown>): FakeGitFile[] {
  const files: FakeGitFile[] = [];
  collectDirectory(root, diffs, files);
  return files;
}

function collectDirectory(
  node: Record<string, unknown> | undefined,
  diffs: Record<string, unknown>,
  files: FakeGitFile[],
): void {
  if (!node) return;
  const nested = Array.isArray(node.directories) ? node.directories : [];
  for (const child of nested) collectDirectory(asRecord(child), diffs, files);
  const listed = Array.isArray(node.files) ? node.files : [];
  for (const item of listed) {
    const file = asRecord(item);
    const path = stringValue(file?.path);
    if (!path) continue;
    const diffNode = asRecord(diffs[path]);
    files.push({
      path,
      status: normalizeGitStatus(stringValue(file?.status) ?? stringValue(diffNode?.status)),
      diff: stringValue(diffNode?.diff) ?? `--- a/${path}\n+++ b/${path}\n`,
    });
  }
}

function ensureIncomingCommit(git: FakeGitState): string {
  git.incomingCommit ??= nextCommitId(git);
  return git.incomingCommit;
}

function nextCommitId(git: FakeGitState): string {
  const id = `f4ke${String(git.nextCommit).padStart(8, '0')}`;
  git.nextCommit += 1;
  return id;
}

function syncResult(git: FakeGitState): FakeGitMutation {
  return { found: true, succeeded: true, error: null, head: gitHeadPayload(git) };
}

function asLogCommit(value: unknown): GitLogCommit | null {
  const record = asRecord(value);
  if (!record) return null;
  const id = stringValue(record.id);
  if (!id) return null;
  return {
    id,
    shortId: stringValue(record.shortId) ?? id.slice(0, 7),
    parents: stringArray(record.parents) ?? [],
    subject: stringValue(record.subject) ?? '',
    author: stringValue(record.author) ?? 'Demo',
    timestamp: stringValue(record.timestamp) ?? new Date().toISOString(),
    refs: stringArray(record.refs) ?? [],
  };
}

function remoteBranches(value: unknown): Array<{ remote: string; name: string }> {
  if (!Array.isArray(value)) return [{ remote: 'origin', name: 'main' }];
  return value
    .map(item => asRecord(item))
    .filter((item): item is Record<string, unknown> => !!item)
    .map(item => ({
      remote: stringValue(item.remote) ?? 'origin',
      name: stringValue(item.name) ?? 'main',
    }));
}

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === 'object' && !Array.isArray(value) ? value as Record<string, unknown> : undefined;
}

function stringValue(value: unknown): string | null {
  return typeof value === 'string' ? value : null;
}

function numberValue(value: unknown): number | null {
  return typeof value === 'number' && Number.isFinite(value) ? value : null;
}

function stringArray(value: unknown): string[] | null {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : null;
}
